### PR TITLE
ARROW-3039: [Go] Add support for DictionaryArray

### PR DESCRIPTION
**Pull request by zeroshade**
Also resolves ARROW-5267, ARROW-7286, and ARROW-9378 which all are different pieces of the DictionaryArray support.

This *does not* implement Dictionary support for scalars yet, nor does it yet support concatenating Dictionary Arrays and dictionary unification. Cards will be created to track work for those two pieces of functionality separately.

### CODE REVIEW

**Comment by brancz** [comment]
Hey! I stumbled on this PR since I was looking for dictionary support, and I was wondering why there wasn't a `StringDictionaryBuilder`, but almost everything else? Or is one just supposed to use the `Dictionary` type directly for that? The other case I have that I'd want to use dictionaries for is Lists, as in a dictionary of common lists seen before, does that even work with Dictionaries? (sorry if the second one is a more general question than about this PR)

**Comment by zeroshade** [comment]
Hey @brancz The reason for a lack of `StringDictionaryBuilder` is that it's really just a special case of the `BinaryDictionaryBuilder` which can be instantiated with the proper `arrow.DataType` and accepts adding either `[]byte` or `string`. So rather than create a separate `StringDictionaryBuilder` I just rolled it into that. Alternatively I could create a `StringDictionaryBuilder` which is just an alias to the `BinaryDictionaryBuilder` if that might be easier to understand?

As for Lists, in the most technical sense nothing in the Arrow format spec prohibits or prevents using a List array as a dictionary. I just simply didn't implement it yet in this PR as that is not a case covered by the integration tests.

I'd love any feedback on the API from real world usage! I'm still waiting on people looking through and reviewing this and so on, and was aiming to not merge this until after arrow version 7.0.0 gets released anyways. So i'd integrate any feedback you have into this PR. Looking forward to it and thanks for checking this out!

**Comment by brancz** [comment]
That makes sense, thank you for elaborating!

File: dev/archery/archery/integration/datagen.py
Comment (emkornfield) [inline dev/archery/archery/integration/datagen.py:1623]: :)

File: go/arrow/array/array_test.go
Comment (emkornfield) [inline go/arrow/array/array_test.go:100]: does it pay to test more index and value types?

File: go/arrow/array/data.go
Comment (emkornfield) [inline go/arrow/array/data.go:39]: maybe expand the comment  to elaborate that for dictionary arrays buffers will contain nullability and indexes  that reference elements in this variable? (and childdata would be empty)

File: go/arrow/array/data.go
Comment (emkornfield) [inline go/arrow/array/data.go:69]: should we check that childData is empty here?

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:36]: Thank you for the nice docs!

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:83]: This needs signed cased  also?

From the spec

```
Since unsigned integers can be more difficult to work with in some cases (e.g. in the JVM), we recommend preferring signed integers over unsigned integers for representing dictionary indices. Additionally, we recommend avoiding using 64-bit unsigned integer indices unless they are required by an application.
```

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:76]: nit: probably will never happen in practice but upperlimit could be awkward here for comparison again int64 in case of overflow (I think I'm OK ignoring it).

File: go/arrow/array/dictionary.go
Comment (emkornfield) [reply go/arrow/array/dictionary.go:83]: I see this is handled below.

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:97]: TODOs with links to JIRAs instead of contributors would be appreciated.

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:93]: couldn't this check be applied to signed values as well?

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:125]: not sure if you care about perf here and what the SIMD version story is in Go but in C++ the compiler can use SIMD to take the min and max over the entire data array first.  Then checking the bounds of that min and max against 0 and upper bound.  (Not sure if the function to do this ended up in the parquet implementation native code).  The down-side is the error message has to be less precise (you can't can provide index without reiterating over the array but most arrays will be valid anyways).

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:263]: might want to clarify that index types also need to be equal

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:291]: unlikely to hit this in practice but doesn't this potentially truncate the index?

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:470]: why is panic the right thing here?

File: go/arrow/array/dictionary.go
Comment (emkornfield) [inline go/arrow/array/dictionary.go:796]: docs?

File: go/arrow/ipc/file_reader.go
Comment (emkornfield) [inline go/arrow/ipc/file_reader.go:165]: might be worth a comment on why update or delta is ignored? (maybe also validate that it is of the expected type).

Comment (emkornfield) [reply go/arrow/ipc/file_reader.go:165]: I might have also missed it but is Kind of dictionary used anywhere?

**Review by emkornfield — CHANGES_REQUESTED** [review]
Biggest concern is over making sure delta and replacement dictionaries are appropriately handled, quickly browsing through the code I couldn't find checks on this.

File: go/arrow/array/data.go
Comment (zeroshade) [reply go/arrow/array/data.go:69]: Is it guaranteed that `childData` should always be empty for Dictionary arrays? Based on the spec, can you have a dictionary of lists where the values of the dictionary are themselves list arrays? If it's guaranteed to always be empty, I'll just remove it from the argument list here.

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:125]: yea, the function to do that natively with SIMD is in the Parquet implemention's native code currently. That's still a good idea for me to lift that into the shared utilities and use that here.

**Comment by emkornfield** [comment]
@zeroshade sorry, I lost track of this, are you waiting for me on this?

**Comment by zeroshade** [comment]
@emkornfield Nope, i'm not waiting for you on this, i was on vacation for a week and then ended up in the hospital with pneumonia. I just haven't had a chance to get back to this to address the comments and fixes. I am gonna try to do so later this week / early next week. I'll tag you in a comment when I'm ready for another lookover by you.

Sorry for the confusion.

**Comment by emkornfield** [comment]
@zeroshade sorry to hear that.  Glad you are on the mend.  I figured I'd check in since I'm usually the bottleneck :)

File: go/arrow/array/data.go
Comment (zeroshade) [reply go/arrow/array/data.go:69]: also, we'd only verify that childData is empty *iff* dict is non-nil.

@emkornfield do you know the answer to whether or not `childData` should *always* be empty for a dictionary array?

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:93]: I've added a comment there as to why we except signed values from this check, but basically:

For unsigned integers, if `upperlimit > maxval` there's no need to bounds check at all, but for signed integers the value could still be < 0, so we still need to do the bounds check to confirm.

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:291]: fair point. On a 64-bit machine if you have uint64 indexes and the value is larger than can fit in an int64 it would truncate. 

On a 32-bit machine this'll truncate for any 64-bit index type and for uint32 indexes that are larger than an int32. I guess I should change this to return a uint64 instead? so that we know we never truncate? What do you think?

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:470]: The alternative would be to change the function signature to return `(DictionaryBuilder, error)` but all of the other functions to construct a NewBuilder currently would panic if they ran into an error and don't return `(Builder, error)` so I was following the existing pattern.

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:241]: Clarified

File: go/arrow/array/data.go
Comment (zeroshade) [reply go/arrow/array/data.go:42]: done

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go]: fixed

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:780]: added

File: go/arrow/ipc/file_reader.go
Comment (zeroshade) [reply go/arrow/ipc/file_reader.go:165]: Good point, i've added the check here so that we error on a replacement since that is not supported for IPC files.

As far as expected type: that is validated inside the `readDictionary` function. We get the expected types from the schema, and read dictionaries using the types as pulled from the memo table populated by the schema. If the type doesn't match, `ctx.loadArray` inside of `readDictionary` will fail with the wrong type.

We don't currently do the stats on the number of dictionaries like the C++ lib does, but it might get added. So usage of the dictionary kind will likely get utilized more than just having the IPC File fail on replacements in a later change.

**Comment by zeroshade** [comment]
@emkornfield Delta and replacement dictionaries are handled at the bottom of the `readDictionary` function by calling `AddDelta` and `AddOrReplace` on the `memo` object, while the `memo` object is used when constructing the record batches with those dictionaries. 

I've added the error case for IPC files to error out on a replacement dictionary. That should cover everything as far as I'm aware for now, and we're passing all the integration tests with dictionaries.

**Comment by zeroshade** [comment]
@emkornfield This is ready for a re-review now! Thanks!

**Comment by zeroshade** [comment]
rebased from master to resolve conflict.

bump @emkornfield for review :smile:

**Comment by zeroshade** [comment]
bump again @emkornfield! 😄

**Comment by emkornfield** [comment]
on my radar, and i should have a little spare time next week.  Sorry for the delay.

File: go/arrow/array/data.go
Comment (emkornfield) [reply go/arrow/array/data.go:72]: the spec allows for arbitrary dictionary values (including lists).  I don't think this is super well supported in the general ecosystem though.

File: go/arrow/array/dictionary.go
Comment (emkornfield) [reply go/arrow/array/dictionary.go:270]: maybe just panic for truncation?  like I said i think this is probably unlikely.  If you want to go through the effort int64 seems sufficient.  most architectures I'm aware of can't even address a full 64 bits of ram.

File: go/arrow/array/dictionary.go
Comment (emkornfield) [reply go/arrow/array/dictionary.go:451]: would it make sense to clean this up at some point later?

**Comment by emkornfield** [comment]
+1 LGTM, mostly reviewed conversation points.

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:270]: Sure, i'll add a check for truncation and panic.

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:451]: probably. 

Honestly, I'd like to replace a lot of this using Go generics that came out with go1.18, but since we want to maintain support for at least `CurrentVersion`-1 I'm going to have to wait until Go1.19 is released before doing so (because switching to Go generics would make this incompatible with any Go versions <1.18).

File: go/arrow/array/dictionary.go
Comment (zeroshade) [reply go/arrow/array/dictionary.go:270]: Actually, given how unlikely this, I'm going to make it a `debug.Assert` to ensure that it doesn't affect performance unless compiled with the asserts turned on.

**Comment by zeroshade** [comment]
Thanks @emkornfield updated based on the comments and I'll merge once this passes the CI

**Pull request closed without merging**
